File: cache/src/main/java/com/osrsd/cache/provider/EnumProvider.java
package com.osrsd.cache.provider;

import com.osrsd.cache.def.Definition;
import com.osrsd.cache.def.EnumDefinition;
import com.osrsd.cache.util.ByteBufferExt;
import lombok.extern.slf4j.Slf4j;

import java.nio.ByteBuffer;
import java.util.stream.IntStream;

@Slf4j
public final class EnumProvider {

    public static Definition decode(ByteBuffer buffer, EnumDefinition definition) {
        do {
            int opcode = buffer.get() & 0xff;
            if (opcode == 0) {
                break;
            }

            if (opcode == 1) {
                definition.setKeyType((char) (buffer.get() & 0xff));
            } else if (opcode == 2) {
                definition.setValType((char) (buffer.get() & 0xff));
            } else if (opcode == 3) {
                definition.setDefaultString(ByteBufferExt.getString(buffer));
            } else if (opcode == 4) {
                definition.setDefaultInt(buffer.getInt());
            } else if (opcode == 5) {
                int size = buffer.getShort() & 0xffff;
                IntStream.range(0, size).map(index -> buffer.getInt()).forEach(key -> {
                    String value = ByteBufferExt.getString(buffer);
                    definition.getValues().put((long) key, value);
                });
            } else if (opcode == 6) {
                int size = buffer.getShort() & 0xffff;
                IntStream.range(0, size).map(index -> buffer.getInt()).forEach(key -> {
                    int value = buffer.getInt();
                    definition.getValues().put((long) key, value);
                });
            } else {
                log.warn(String.format("Unhandled definition opcode with id: %s.", opcode));
            }
        } while (true);
        return definition;
    }

}

File: app/src/main/java/com/osrsd/cache/provider/EnumProvider.kt
package com.osrsd.cache.provider

import com.osrsd.spring.domain.Definition
import com.osrsd.cache.util.ByteBufferExt
import com.osrsd.spring.domain.EnumDefinition
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer

object EnumProvider {

    private val log: Logger = LoggerFactory.getLogger(EnumProvider::class.java)

    fun decode(buffer: ByteBuffer, definition: EnumDefinition): Definition {
        do {
            when (val opcode: Int = buffer.get().toInt() and 0xff) {
                1 -> definition.keyType = (buffer.get().toInt() and 0xff).toChar()
                2 -> definition.valType = (buffer.get().toInt() and 0xff).toChar()
                3 -> definition.defaultString = ByteBufferExt.getString(buffer)
                4 -> definition.defaultInt = buffer.int
                5 -> {
                    val length: Int = buffer.short.toInt() and 0xffff
                    (0 until length).forEach { _ ->
                        val key = buffer.int
                        val value = ByteBufferExt.getString(buffer)
                        definition.params[key.toLong()] = value
                    }
                }
                6 -> {
                    val length: Int = buffer.short.toInt() and 0xffff
                    (0 until length).forEach { _ ->
                        val key = buffer.int
                        val value: Int = buffer.int
                        definition.params[key.toLong()] = value.toString()
                    }
                }
                0 -> break
                else -> log.warn(String.format("Unhandled definition opcode with id: %s.", opcode))
            }
        } while (true)
        return definition
    }

}
